File: py/medium/DailyTemperatures.py
from typing import List


class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        size = len(temperatures)
        ans, st = [0] * size, [0]
        for i in range(size):
            while st and temperatures[i] > temperatures[st[-1]]:
                idx = st.pop()
                ans[idx] = i - idx
            st.append(i)

        return ans


s = Solution()
print(s.dailyTemperatures([30, 38, 30, 36, 35, 40, 28]))

File: ts/medium/DailyTemperatures.js
class Solution {
	/**
	 * @param {number[]} temperatures
	 * @return {number[]}
	 */
	dailyTemperatures(temperatures) {
		const stack = [];
		const result = new Array(temperatures.length).fill(0);

		for (let idx = 0; idx < temperatures.length; idx++) {
			const currentTemp = temperatures[idx];
			while (stack.length !== 0 && currentTemp > stack[stack.length - 1][0]) {
				const [_, stackIdx] = stack.pop()
				const days = idx - stackIdx;
				result[stackIdx] = days;
			}
			stack.push([currentTemp, idx]);
		}
		return result;
	}
}
